str/fine-mapping/finemap_files_prep.py

```diff
@@ -0,0 +1,151 @@
+    df = df.reindex(df_index['Unnamed: 0'])
+
+    # Reset the index of df
+    df = df.reset_index()
+    df = df.rename(columns={'Unnamed: 0': 'rsid'})
+
+    output_file_path = output_path(f'finemap_prep/{celltype}/{chrom}/{gene_name}.z', 'analysis')
+    # Save the DataFrame to a space-delimited file
+    df.to_csv(output_file_path, sep=' ', index=False)
+
+
+def ld_file_maker(gene_name, ld_file, celltype, chrom):
+    """
+    Prepares the 'ld file' for FINEMAP based on the output from corr_matrix_maker.py.
+    """
+
+    # Load the TSV file
+    df = pd.read_csv(ld_file, sep='\t')
+    df = df.drop(columns=['Unnamed: 0'])
+
+    # Convert the data type of the DataFrame to np.float64
+    df = df.astype(np.float64)
+
+    # Rounding to prevent float representation errors by FINEMAP
+    df = df.round(4)
+
+    # Remove column names
+    df.columns = ['' for _ in df.columns]
+
+    # Remove index name
+    df.index.names = [None]
+
+    output_file_path = output_path(f'finemap_prep/{celltype}/{chrom}/{gene_name}.ld', 'analysis')
+    # Save the DataFrame to a space-delimited file without index
+    df.to_csv(output_file_path, sep=' ', index=False, header=False)
+
+
+@click.option('--ld-dir', required=True, help='Path to the directory containing the LD files')
+@click.option('--associatr-dir', required=True, help='Path to the directory containing the associatr files')
+@click.option('--celltypes', required=True, help='Cell type')
+@click.option('--chroms', required=True, help='Chromosome')
+@click.option('--ld-job-cpu', help='CPU for LD job', default=0.25)
+@click.option('--z-job-cpu', help='CPU for Z job', default=0.25)
+@click.option('--max-parallel-jobs', help='Maximum number of parallel jobs to run', default=500)
+@click.command()
+def main(ld_dir, associatr_dir, celltypes, chroms, ld_job_cpu, z_job_cpu, max_parallel_jobs):
+    # Setup MAX concurrency by genes
+    _dependent_jobs: list[hb.batch.job.Job] = []
+
+    def manage_concurrency_for_job(job: hb.batch.job.Job):
+        """
+        To avoid having too many jobs running at once, we have to limit concurrency.
+        """
+        if len(_dependent_jobs) >= max_parallel_jobs:
+            job.depends_on(_dependent_jobs[-max_parallel_jobs])
+        _dependent_jobs.append(job)
+
+    b = get_batch(name=f'FINEMAP prep for {celltypes}: {chroms}')
+    for celltype in celltypes.split(','):
+        for chrom in chroms.split(','):
+            ld_files = list(to_path(f'{ld_dir}/{celltype}/{chrom}').glob('*.tsv'))
+            for ld_file in ld_files:
+                gene_name = str(ld_file).split('/')[-1].split('_')[0]  # ENSG ID
+
+                if (
+                    to_path(
+                        output_path(f'finemap_prep/{celltype}/{chrom}/{gene_name}.ld', 'analysis'),
+                    ).exists()
+                    and to_path(output_path(f'finemap_prep/{celltype}/{chrom}/{gene_name}.z', 'analysis')).exists()
+                ):
+                    continue
```

Comment:
not a biggie, but be aware of your job scaling (type * chrom * gene), and you're creating 2 existence checks per combination. All that happens in the driver job so you're delaying the start of the actual work.

I'd experiment with a change here - 

```
# get all files in the output folder, recursively, in a single query
all_files = list(to_path(output_path('finemap_prep', 'analysis')))

# check whether your intended outputs are in that list
ld_file = output_path(f'finemap_prep/{celltype}/{chrom}/{gene_name}.ld', 'analysis')
z_file = output_path(f'finemap_prep/{celltype}/{chrom}/{gene_name}.z', 'analysis')
if all (filepath in all_files for filepath in [z_file, ld_file]):
    continue
```

I thiiiiink this should scale a lot better, by posting one large query instead of thousands of individual ones.

This also builds the full output file names, so you can pass them to the relevant methods (you pass the celltype, chrom, and gene name to your methods, but you already made the full path here to check if it exists)